Mathlib/Analysis/Normed/Ring/SmoothingSeminorm.lean

```diff
@@ -0,0 +1,604 @@
+
+/-!
+# smoothingSeminorm
+In this file, we prove [BGR, Proposition 1.3.2/1][bosch-guntzer-remmert] : if `f` is a
+nonarchimedean seminorm on `R`, then `iInf (fun (n : pnat), (f(x^(n : ℕ)))^(1/(n : ℝ)))` is a
+power-multiplicative nonarchimedean seminorm on `R`.
+
+## Main Definitions
+* `smoothingSeminormSeq` : the `ℝ`-valued sequence sending `n` to `(f (x ^ n))^(1/n : ℝ)`.
+* `smoothingSeminorm'` : the iInf of the sequence `f(x^(n : ℕ)))^(1/(n : ℝ)`.
+* `smoothingSeminorm` : if `f 1 ≤ 1` and `f` is nonarchimedean, then `smoothingSeminorm'`
+  is a ring seminorm.
+
+## Main Results
+
+* `tendsto_smoothingSeminorm'` :if `f 1 ≤ 1`, then `smoothingSeminorm' f x` is the limit
+  of `smoothingSeminormSeq f x` as `n` tends to infinity.
+* `isNonarchimedean_smoothingSeminorm'` : if `f 1 ≤ 1` and `f` is nonarchimedean, then
+  `smoothingSeminorm'` is nonarchimedean.
+* `isPowMul_smoothingSeminorm'` : if `f 1 ≤ 1` and `f` is nonarchimedean, then
+  `smoothingSeminorm' f` is power-multiplicative.
+
+## References
+* [S. Bosch, U. Güntzer, R. Remmert, *Non-Archimedean Analysis*][bosch-guntzer-remmert]
+
+## Tags
+
+smoothingSeminorm, seminorm, nonarchimedean
+-/
+
+noncomputable section
+
+open Filter Nat Real
+
+open scoped Topology NNReal
+
+variable {R : Type*} [CommRing R] (f : RingSeminorm R)
+
+section smoothingSeminorm
+
+/-- The `ℝ`-valued sequence sending `n` to `(f (x ^ n))^(1/n : ℝ)`. -/
+def smoothingSeminormSeq (x : R) : ℕ → ℝ := fun n => f (x ^ n) ^ (1 / n : ℝ)
+
+/-- For any positive `ε`, there exists a positive natural number `m` such that
+  `f (x ^ (m : ℕ)) ^ (1 / m : ℝ) < iInf (fun (n : pnat), (f(x ^(n : ℕ)))^(1/(n : ℝ))) + ε/2`. -/
+private theorem smoothingSeminormSeq_hasLimit_m (x : R) {ε : ℝ} (hε : 0 < ε) :
+    ∃ m : PNat, f (x ^ (m : ℕ)) ^ (1 / m : ℝ) <
+        (iInf fun n : PNat => f (x ^ (n : ℕ)) ^ (1 / (n : ℝ))) + ε / 2 :=
+  exists_lt_of_ciInf_lt (lt_add_of_le_of_pos (le_refl _) (half_pos hε))
+
+private theorem smoothingSeminormSeq_hasLimit_aux {L : ℝ} (hL : 0 ≤ L) {ε : ℝ} (hε : 0 < ε)
+    {m1 : ℕ} (hm1 : 0 < m1) {x : R} (hx : f x ≠ 0) :
+    Tendsto
+      (fun n : ℕ => (L + ε) ^ (-(((n % m1 : ℕ) : ℝ) / (n : ℝ))) * (f x ^ (n % m1)) ^ (1 / (n : ℝ)))
+      atTop (𝓝 1) := by
+  rw [← mul_one (1 : ℝ)]
+  have h_exp : Tendsto (fun n : ℕ => ((n % m1 : ℕ) : ℝ) / (n : ℝ)) atTop (𝓝 0) :=
+    tendsto_mod_div_atTop_nhds_zero_nat hm1
+  apply Tendsto.mul
+  · have h0 : Tendsto (fun t : ℕ => -(((t % m1 : ℕ) : ℝ) / (t : ℝ))) atTop (𝓝 0) := by
+      rw [← neg_zero]
+      exact Tendsto.neg h_exp
+    rw [← rpow_zero (L + ε)]
+    apply Tendsto.rpow tendsto_const_nhds h0
+    rw [ne_eq, add_eq_zero_iff_of_nonneg hL (le_of_lt hε)]
+    exact Or.inl (not_and_of_not_right _ (ne_of_gt hε))
+  · simp_rw [mul_one, ← rpow_natCast, ← rpow_mul (apply_nonneg f x), ← mul_div_assoc, mul_one, ←
+      rpow_zero (f x)]
+    exact Tendsto.rpow tendsto_const_nhds h_exp (Or.inl hx)
+
+/-- `smoothingSeminormSeq` is bounded below by zero. -/
```

Comment:
The statement only says it is bounded below; do you want to add a lemma saying that the bound is `0`?

Reply:
I only need that it is bounded below, but I could add that lemma if you think it would be helpful.

Reply:
Yes, this might be a good idea.